### ISSUE TITLE
IDL: typedef causes problems with keyfield

### ISSUE THREAD

**Issue by damiandixon** [eclipse-cyclonedds/cyclonedds-cxx#29]
Ubuntu 20.04

The following IDL works with opensplice but fails with the current C++ idl compiler of cyclonedds:
```
  enum e_type
  {
     EE_A,
     EE_L,
     EE_S1,
     EE_S2,
     EE_S3
  };

  typedef struct tag_Ident_T {
    long id;
    e_type idType;
  } Ident_T;

  struct Report_T {
    Ident_T    id;    //@key
    long       value;
  };
  #pragma keylist Report_T id.id id.idType value
```
Failure messages:
```
HelloWorldData.idl:34, keyfield id.id is either missing or unsupported
HelloWorldData.idl:34, keyfield id.idType is either missing or unsupported
```
The work around is to remove the typedef and tag_Ident_T. However I've too many definitions supplied to me for me to get away with making this change.

**Comment by eboasson** [comment]
https://github.com/ADLINK-IST/idlpp-cxx/pull/6 solves this particular problem for me, but it might introduce new problems ... I'd be much obliged if you would check.

I suppose that I may infer from your writing "current C++ idl" that you are aware that we're (finally!) sunsetting this abomination of an IDL compiler. Unfortunately, I think it is still a bit too early for C++ users to switch to it and besides the C backend doesn't yet exist, and that means "the abomination" will be with us a little while longer. I'll try to keep it from being a showstopper.

**Comment by damiandixon** [comment]
The IDL compiles now.

However linking fails:
```
[ 46%] Linking CXX executable ddscxxHelloworldSubscriber
/usr/bin/ld: CMakeFiles/ddscxxHelloworldSubscriber.dir/HelloWorldDataSplDcps.cpp.o: in function `__HelloWorldData_Report_T__copyIn(void const*, void*)':
HelloWorldDataSplDcps.cpp:(.text+0xb4): undefined reference to `__HelloWorldData_Ident_T__copyIn(HelloWorldData::tag_Ident_T const*, HelloWorldData_tag_Ident_T*)'
collect2: error: ld returned 1 exit status
```
Commenting out the `extern void __HelloWorldData_Ident_T__copyIn(const ::HelloWorldData::Ident_T *, HelloWorldData_Ident_T *);` in the generated file HelloWorldDataSplDcps.cpp allows the compilation to complete. See attached txt file line 44.

[HelloWorldDataSplDcps.txt](https://github.com/eclipse-cyclonedds/cyclonedds-cxx/files/5327987/HelloWorldDataSplDcps.txt)

> I suppose that I may infer from your writing "current C++ idl" 
I'm aware the current IDL compiler is being replaced at somepoint.
I just thought I would see how far I could get.

**Comment by damiandixon** [comment]
When I try to compile my large project I still have this particular IDL error.

I'll have to look at trying to create a slightly different problem example.

The jist is the typedef is in a different module and file from where it is used.

**Comment by eboasson** [comment]
> Commenting out the extern void __HelloWorldData_Ident_T__copyIn(const ::HelloWorldData::Ident_T *, HelloWorldData_Ident_T *); in the generated file HelloWorldDataSplDcps.cpp allows the compilation to complete. See attached txt file line 44.

I think the below patch achieves pretty much the same effect:
```diff
diff --git a/idlpp/ppresources/ospli/src/tools/idlpp/code/idl_genLiteISOCxxCopyin.c b/idlpp/ppresources/ospli/src/tools/idlpp/code/idl_genLiteISOCxxCopyin.c
index d0343a0..f4cc72d 100644
--- a/idlpp/ppresources/ospli/src/tools/idlpp/code/idl_genLiteISOCxxCopyin.c
+++ b/idlpp/ppresources/ospli/src/tools/idlpp/code/idl_genLiteISOCxxCopyin.c
@@ -359,8 +359,8 @@ idl_structureMemberOpenClose (
             IDL_PRINTLINE (indent);
             idl_fileOutPrintf (idl_fileCur(), "        extern void __%s__copyIn(const %s *, %s *);\n",
                         scopedTypeName,
-                    scopedCxxTypeName,
-                    scopedTypeName);
+                        "void"/*scopedCxxTypeName*/,
+                        "void"/*scopedTypeName*/);
             idl_fileOutPrintf (idl_fileCur(), "        __%s__copyIn(&from->%s(), &to->%s);\n",
                         scopedTypeName,
                     cid,
```

I really don't understand why there're `void *`s all over the place when proper type information is actually available ... but whatever.

> The jist is the typedef is in a different module and file from where it is used.

Is a bigger problem. These typedef issues come from the Java-based IDL compiler that Cyclone DDS inherited and that is also used for generating C code. For C++, perhaps the new one is serviceable already (a medium shot, but you're welcome to try).

My earlier modification simply called `stresolve` to do a name lookup, but that uses the scope in effect at the point of use. It therefore only works if the names are all defined in the same scope. When you move the typedef into a different module the first name lookup finds the typedef starting from the "current" scope, but then attemps to resolve the typedef in that same "current scope". And that fails.

If I only I knew exactly what it was doing and how to get the scope in which to perform the second lookup ... I'll give it another bit of thought.

**Comment by damiandixon** [comment]
The above patch works for the simple case of everything in the same file. Thanks.

I'll give the idlcxx branch ago.

**Comment by eboasson** [comment]
It looks like the entries in the symbol table hold fully scoped names, so taking the name under which the typedef is stored in the symbol table as the starting point for the next lookup seems like it should do the trick. See https://github.com/ADLINK-IST/idlpp-cxx/pull/6 

It does allow me to compile with my mangled HelloWorldData.idl:

```
/*
 * Copyright(c) 2006 to 2020 ADLINK Technology Limited and others
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
 * v. 1.0 which is available at
 * http://www.eclipse.org/org/documents/edl-v10.php.
 *
 * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
 */
module XYZ
{
  enum e_type
  {
     EE_A,
     EE_L,
     EE_S1,
     EE_S2,
     EE_S3
  };
  typedef struct tag_Ident_T {
    long id;
    e_type idType;
  } Ident_T;
};

module HelloWorldData
{
  struct Report_T {
    XYZ::Ident_T    id;    //@key
    long       value;
  };
  #pragma keylist Report_T id.id id.idType value
  union UU switch (boolean) { case TRUE: Report_T v; }; 
  struct Msg
  {
    long userID;
    string message;
    UU uu;
  };
  #pragma keylist Msg userID
};
```

**Comment by eboasson** [comment]
@damiandixon on the assumption that the positive result I get means the problem is now also fixed when the types are different modules, we merged the changes (and as you can see I have also opened a PR for the main repository so that those using C will get the fixes as well).

It would be great if you could check whether it really does solve the problem for your set of IDL files. If it does, I can close the ticket.

**Comment by damiandixon** [comment]
@eboasson the update almost fixes this problem.

Sorry I should have created an example which more closely represented how my IDL was structured.

I've created a new example that demonstrates the new(ish) but related issue.

[helloworld.zip](https://github.com/eclipse-cyclonedds/cyclonedds-cxx/files/5341059/helloworld.zip)

Compiler output:
```
[ 13%] Building CXX object CMakeFiles/ddscxxHelloworldSubscriber.dir/subscriber.cpp.o
In file included from /home/damian/projects/dependencies/cyclone-dds/cyclonedds-cxx/examples/helloworld/build/HelloWorldDataSplDcps.h:1,
                 from /home/damian/projects/dependencies/cyclone-dds/cyclonedds-cxx/examples/helloworld/build/HelloWorldData_DCPS.hpp:7,
                 from /home/damian/projects/dependencies/cyclone-dds/cyclonedds-cxx/examples/helloworld/subscriber.cpp:21:
/home/damian/projects/dependencies/cyclone-dds/cyclonedds-cxx/examples/helloworld/build/HelloWorldData-cyclone.h:21:16: error: redefinition of ‘struct Types_tag_Ident_T’
   21 | typedef struct Types_tag_Ident_T
      |                ^~~~~~~~~~~~~~~~~
In file included from /home/damian/projects/dependencies/cyclone-dds/cyclonedds-cxx/examples/helloworld/build/HelloWorldData-cyclone.h:10,
                 from /home/damian/projects/dependencies/cyclone-dds/cyclonedds-cxx/examples/helloworld/build/HelloWorldDataSplDcps.h:1,
                 from /home/damian/projects/dependencies/cyclone-dds/cyclonedds-cxx/examples/helloworld/build/HelloWorldData_DCPS.hpp:7,
                 from /home/damian/projects/dependencies/cyclone-dds/cyclonedds-cxx/examples/helloworld/subscriber.cpp:21:
/home/damian/projects/dependencies/cyclone-dds/cyclonedds-cxx/examples/helloworld/build/Types-cyclone.h:32:16: note: previous definition of ‘struct Types_tag_Ident_T’
   32 | typedef struct Types_tag_Ident_T
      |                ^~~~~~~~~~~~~~~~~
In file included from /home/damian/projects/dependencies/cyclone-dds/cyclonedds-cxx/examples/helloworld/build/HelloWorldDataSplDcps.h:1,
                 from /home/damian/projects/dependencies/cyclone-dds/cyclonedds-cxx/examples/helloworld/build/HelloWorldData_DCPS.hpp:7,
                 from /home/damian/projects/dependencies/cyclone-dds/cyclonedds-cxx/examples/helloworld/subscriber.cpp:21:
/home/damian/projects/dependencies/cyclone-dds/cyclonedds-cxx/examples/helloworld/build/HelloWorldData-cyclone.h:25:3: error: conflicting declaration ‘typedef int Types_tag_Ident_T’
   25 | } Types_tag_Ident_T;
      |   ^~~~~~~~~~~~~~~~~
In file included from /home/damian/projects/dependencies/cyclone-dds/cyclonedds-cxx/examples/helloworld/build/HelloWorldData-cyclone.h:10,
                 from /home/damian/projects/dependencies/cyclone-dds/cyclonedds-cxx/examples/helloworld/build/HelloWorldDataSplDcps.h:1,
                 from /home/damian/projects/dependencies/cyclone-dds/cyclonedds-cxx/examples/helloworld/build/HelloWorldData_DCPS.hpp:7,
                 from /home/damian/projects/dependencies/cyclone-dds/cyclonedds-cxx/examples/helloworld/subscriber.cpp:21:
/home/damian/projects/dependencies/cyclone-dds/cyclonedds-cxx/examples/helloworld/build/Types-cyclone.h:36:3: note: previous declaration as ‘typedef struct Types_tag_Ident_T Types_tag_Ident_T’
   36 | } Types_tag_Ident_T;
      |   ^~~~~~~~~~~~~~~~~

```
Essentially the struct is declared in full in two different files.

I did try idlcxx branch but I did not have much luck. I suspect some of that may be down to user error on behalf. I'll have to try again when I have a little bit more time.

**Comment by eboasson** [comment]
@damiandixon I don't think I fully understand why it fails in exactly that way: it requires writing `typedef struct X { } Y;` and then using `Y`, splitting it into definitions makes the problem go away! Methinks it is silly to have such a dependency in a code generator on the grammatical construct used in the source ...

What I did notice is that the compiler guards the generation of output for a given definition by whether or not the definition is in the top-level IDL file, but then makes an exception for `enum`s and `struct`s that are (in some sense) reached via a member. That exception allows the definition of the tagged struct to be output. I've thought about it a bit and I don't see how the exception could possibly be correct.

So I have made a PR. I hope you're willing to give it a try: https://github.com/ADLINK-IST/idlpp-cxx/pull/7

**Comment by damiandixon** [comment]
@eboasson The example builds and links.

My project code builds but fails to link.

It looks like the change to idl_genLiteISOCxxCopyin.c maybe incomplete as some of the link failures are around the functions being called. Problem is I can't say definitively that the idl compiler is at fault as I'm having fun with my own projects build configuration for cyclone-dds hence the need to find the for time to play.

**Comment by damiandixon** [comment]
@eboasson if the typedef is embeded in another model then the code generated fails to link.

```
/usr/bin/ld: CMakeFiles/ddscxxHelloworldSubscriber.dir/HelloWorldDataSplDcps.cpp.o: in function `__HelloWorldData_Report_T__copyIn(void const*, void*)':
HelloWorldDataSplDcps.cpp:(.text+0x3c): undefined reference to `__Types_common_Ident_T__copyIn(Types::common::tag_Ident_T const*, Types_common_tag_Ident_T*)'
collect2: error: ld returned 1 exit status
```
The idl is as follows:
```
#include "Types.idl"

module HelloWorldData
{
  struct Report_T {
    Types::common::Ident_T    id;    //@key
    long       value;
  };
  #pragma keylist Report_T id.id id.idType value
};
```
And `Types.idl`:
```
module Types
{
  module common
  {
    enum e_type
    {
       EE_A,
       EE_L,
       EE_S1,
       EE_S2,
       EE_S3
    };

    typedef struct tag_Ident_T {
      long id;
      e_type idType;
    } Ident_T;
  };
};
```
I've attached an updated example.

[multiple_level_modules.zip](https://github.com/eclipse-cyclonedds/cyclonedds-cxx/files/5359114/multiple_level_modules.zip)

The issue is that the `__Types_common_Ident_T__copyIn` method is implemented with `void*` parameters in one file but used with the actual types in another file.

**Comment by damiandixon** [comment]
Not sure if you want this one as a new issue. It is however similar.

Linker failure.
```
/usr/bin/ld: CMakeFiles/ddscxxHelloworldSubscriber.dir/HelloWorldDataSplDcps.cpp.o: in function `__HelloWorldData_Report_T__copyIn(void const*, void*)':
HelloWorldDataSplDcps.cpp:(.text+0x166): undefined reference to `__HelloWorldData_DataList__copyIn(void const*, void*)'
```

idl:
```
module HelloWorldData
{
  struct Data {
    boolean avail;
  };

  typedef sequence<Data,20> DataList;

  struct Report_T {
    long       value;
    DataList   dataList;
  };
  #pragma keylist Report_T value

  struct Msg
  {
    long userID;
    string message;
  };
  #pragma keylist Msg userID
};
```

**Comment by eboasson** [comment]
Indeed, it is similar. What a mess! I'm glad that IDL compiler's days are numbered ...

I've made a pull request that always erases all types on these "copy in" functions: https://github.com/ADLINK-IST/idlpp-cxx/pull/9 . I hope that's the last hurdle!

**Comment by damiandixon** [comment]
@eboasson that seems to fix the last problem. Thanks for your help.

All the IDL I have now seems to generate and compile.

The issues I have left I believe are down to my build system not linking correctly or to do with the use of the library.

**Comment by k0ekk0ek** [comment]
I think we can safely close this issue. New IDL compiler will be used on the master branch shortly once the *templated-streaming* branch is merged. The new serializer introduced by @reicheratwork also improves performance quite a bit.
